lib/packwerk/const_node_inspector.rb

```diff
@@ -9,12 +9,12 @@ class ConstNodeInspector
     include ConstantNameInspector
 
     def constant_name_from_node(node, ancestors:)
-      return nil unless Node.type(node) == Node::CONSTANT
+      return nil unless Node.constant?(node)
```

Comment:
There’s a fair amount of discussion about this design decision on the (Shopify-only) Shopify/packwerk-old#266 PR. The idea is that there were many places in Packwerk where we wanted to test for multiple node types, e.g. “if this is a class or module or constant assignment, then …, otherwise …”, so the API was designed to be usable in a `case` statement:

```ruby
case Node.type(node)
when Node::CLASS, Node::MODULE, Node::CONSTANT_ASSIGNMENT
  # …
when Node::BLOCK
  # …
else
  # …
end
```

If we provide predicates instead then this becomes a slightly less convenient `if Node.class?(node) || Node.module?(node) || Node.constant_assignment?(node)`. I was reluctant to provide both because that would mean our `Packwerk::Node` API offers two different ways of doing the same thing, which makes it (again slightly) less straightforward to use.

Looking at the code now, I can see that essentially all of these `case` statements have been moved into `Packwerk::Node` itself, so perhaps there’s a case for making `Node.type` private and exposing only predicates to the rest of Packwerk. WDYT?

Reply:
> API offers two different ways of doing the same thing

Agreed, I think it makes sense to keep the API as straightforward to use as possible.

> Looking at the code now, I can see that essentially all of these case statements have been moved into Packwerk::Node itself, so perhaps there’s a case for making Node.type private and exposing only predicates to the rest of Packwerk. 

I did a quick [experiment to verify that we use it only in `Packwerk::Node`](https://github.com/Shopify/packwerk/pull/55/commits/84746c5b4d7772e869311ceb5b380bd527ca8d15) and we do. So it shouldn't be an issue to change `Node.type` to be private.

Looking at the Node Module and how it uses the switch statements it seems to me that we are missing a Factory here. What I would suggest for a separate PR is to move as much logic out of the Node module into dedicated modules converting the Node module into a Factory module. And then as you mentioned change the `Node.type` method to be private. I think this would clean up the code and simplify the API of `Packwerk::Node`. WDYT?

Reply:
> Looking at the Node Module and how it uses the switch statements, it seems to me that we are missing a Factory here. I would suggest a separate PR to move as much logic out of the Node module into dedicated modules, converting the Node module into a Factory module.

When we discussed this on a call, I think we agreed that a) it’s a good idea but b) you’re not going to do it at the moment.